test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+  var invalidCredentials = {email: 'foo1@bar.com', password: 'invalid'};
+  var incompleteCredentials = {password: 'bar1'};
+
+  var validCredentialsUser = null;
+  var validCredentialsUser2 = null;
+
+  // Create a local app variable to prevent clashes with the global
+  // variable shared by all tests. While this should not be necessary if
+  // the tests were written correctly, it turns out that's not the case :(
+  var app;
+
+  beforeEach(function setupAppAndModels(done) {
+    // override the global app object provided by test/support.js
+    // and create a local one that does not share state with other tests
+    app = loopback({localRegistry: true, loadBuiltinModels: true});
+    app.dataSource('db', {connector: 'memory'});
+
+    // attach User and related models
+    User = app.registry.createModel({
+      name: 'TestUser',
+      base: 'User',
+      // forceId is set to false for the purpose of updating the same affected user within the
+      // `Email Update` test cases.
+      forceId: false,
+      // Speed up the password hashing algorithm for tests
+      saltWorkFactor: 4,
+    });
+    app.model(User, {dataSource: 'db'});
+
+    // create and attach a second User-based model
+    User2 = app.registry.createModel({
+      name: 'TestUser2',
+      base: 'User',
+      // forceId is set to false for the purpose of updating the same affected user within the
+      // `Email Update` test cases.
+      forceId: false,
+      // Speed up the password hashing algorithm for tests
+      saltWorkFactor: 4,
+    });
+    app.model(User2, {dataSource: 'db'});
+
+    AccessToken = app.registry.getModel('AccessToken');
+    app.model(AccessToken, {dataSource: 'db'});
+
+    // Update AccessToken and Users relations to use polymorphism
+    AccessToken.belongsTo('user', {idName: 'id', polymorphic: {idType: 'string',
+      foreignKey: 'userId', discriminator: 'principalType'}});
+    User.hasMany('accessTokens', {polymorphic: {foreignKey: 'userId',
+      discriminator: 'principalType'}});
+    User2.hasMany('accessTokens', {polymorphic: {foreignKey: 'userId',
+      discriminator: 'principalType'}});
+
+    // Speed up the password hashing algorithm
+    // for tests using the built-in User model
+    User.settings.saltWorkFactor = 4;
+
+    // allow many User.afterRemote's to be called
+    User.setMaxListeners(0);
+
+    app.enableAuth({dataSource: 'db'});
+    app.use(loopback.token({model: AccessToken}));
+    app.use(loopback.rest());
+
+    User.create(validCredentials, function(err, user) {
+      if (err) return done(err);
+      validCredentialsUser = user;
+      User2.create(validCredentials, function(err, user2) {
+        if (err) return done(err);
+        validCredentialsUser2 = user2;
+        done(null, user, user2);
+      });
+    });
+  });
+
+  describe('User.login', function() {
+    it('Login a user from 1st user model by providing credentials', function(done) {
```

Comment:
The idea behind BDD interface is to structure the tests as a readable specification. The `it` stands for the subject of the outer `describe`. Here, you get the following sentence which looks weird to me:

```
User.login Login a user from 1st user model ...
```

It's better to start the test name with a verb, for example:

```js
it('logs in a user from 1st user model...')
```

The same comment applies to most (if not all) test names below, please revise them too.